### PR TITLE
Feat: Add in ability to delete a single output by prompt name in editor

### PR DESCRIPTION
**Pull request by justinkchen**
https://github.com/lastmile-ai/aiconfig/issues/1451
Screenshot 2024-03-17 at 4 37 15 PM

Adds the ability to delete a single output in the config editor

### CODE REVIEW

File: python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx
Comment (saqadri) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx:208]: Instead of adding the button here, can we add it to the prompt menu?

https://github.com/lastmile-ai/aiconfig/blob/main/python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx

I imagine we will have other capabilities like Copy/Duplicate prompt which will also be added in the future, so it will prevent too many buttons everywhere

**Review by saqadri — COMMENTED** [review]
Looks great overall!! Nice first diff 😃 -- just one suggestion about the button moved to https://github.com/lastmile-ai/aiconfig/blob/main/python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (saqadri) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:59]: nit: clear output may not need to be red, but unsure.

**Review by saqadri — APPROVED** [review]
Ship it! 🚀🚀🚀

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (saqadri) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:59]: Also Delete Prompt should probably remain the last option in the menu.

File: python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:72]: nit: we probably don't need to pass in `promptId` anymore if passing in prompt

File: python/src/aiconfig/editor/client/src/reducers/aiconfigReducer.ts
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/reducers/aiconfigReducer.ts:163]: We have a utils reducer method called `reduceReplacePrompt()`. Would prefer to use that to keep things cleaner and more abstracted

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:33]: This is a front-end thing that Ryan taught me. Since `PromptMenuButton` is already wrapped with a `memo` component, it's actually more overhead for us to wrap these sub-functions in a `useCallback()` hook. Here `useCallback()` checks for changes to any dependencies, but in our case that's simply changes to the props, which is already memoized in the `PromptMenuButton` component itself, so in this case it adds overhead. Would be better to either:

1. Keep inlined
2. Keep separate but don't wrap with `useCallback()`

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:54]: cc @rholinshead for whether this would work with light mode theming

File: python/src/aiconfig/editor/server/server.py
Comment (rossdanlm) [inline python/src/aiconfig/editor/server/server.py:815]: nice thanks!

Comment (rossdanlm) [reply python/src/aiconfig/editor/server/server.py:815]: Actually upon further inspection, I think this should just be `method_name`, not `f"method_{method_name}"`: https://github.com/lastmile-ai/aiconfig/blob/f0f052231feee5f15843e2d88a70f7b406527dd5/python/src/aiconfig/editor/server/server_utils.py#L501-L503

File: python/src/aiconfig/editor/server/server.py
Comment (rossdanlm) [inline python/src/aiconfig/editor/server/server.py:837]: Oh cool, TIL that we already have existing API for this: https://github.com/lastmile-ai/aiconfig/blob/f0f052231feee5f15843e2d88a70f7b406527dd5/python/src/aiconfig/schema.py#L1005. Thought it was for deleting entire outputs only. We can probly clean this up and delete the line `existing_outputs = prompt.outputs`. cc @Ankush-lastmile do you remember why this line might have been needed?

**Review by rossdanlm — APPROVED** [review]
Besides the `useCallbacks` and potential light mode issue, there are no functional blockers I see on my end. I'm happy to unblock and land for now and we can fix lingering nits in a followup! Let me know whether you prefer to fix these small things in this PR (will wait for you) or do a followup (will merge this for now)

File: python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx
Comment (rholinshead) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx:193]: What does this look like / is this needed now that the button is moved to the menu?

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (rholinshead) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:19]: Passing `prompt` through just to access the output length will mean this component is re-rendered any time any part of the prompt changes (e.g. if typing in the input). That will result in the React memo constantly re-calculating the diff to check if it should re-render, and also re-rendering unnecessarily. Instead, it would be better to have a prop `showDeleteOutput` which is set to true if the prompt.outputs.length > 0 in the parent component

**Comment by rossdanlm** [comment]
Btw for failing blockers (on autoformatting), you can use either of these commands to reformat and get unblocked (I think, not 100% sure):

1. `black --exclude '/\.sl\b' . --line-length=79`
2. `fd --glob '*.py'  python/src/aiconfig/editor | xargs python -m 'scripts.lint' --mode=fix --files`

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (rholinshead) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:33]: Oh just noticed as well but we should just construct the `onDeletePrompt` and `onDeleteOutput` callbacks in the parent, since the promptId is known there. Then `promptId` wouldn't even need to be passed, just the callbacks. Similarly that would make PromptMenuButton memoization actually work as intended since the inline function currently prevents memoization from working (new function prop each time). So in `PromptsContainer` we could have
```
const promptId = prompt._ui.id;

const onDeletePrompt = props.onDeletePrompt;
const onDeleteOutput = props.onDeleteOutput;

const deletePrompt = useCallback(() => onDeletePrompt(promptId), [onDeletePrompt, promptId]);
const deleteOutput = useCallback(() => onDeleteOutput(promptId), [onDeleteOutput, promptId]);
```
And they can be passed through the props to PromptMenuButton

File: python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx
Comment (rholinshead) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptMenuButton.tsx:54]: Good catch. I think we can remove color here and it will have the correct color we want. To test this, you can set `mode="gradio"` and `themeMode="dark"` and `themeMode="light"` props for AIConfigEditor in the `LocalEditor` component

File: python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx
Comment (rholinshead) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:38]: This can be removed as unused, since it's already obtained from context.

File: python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx
Comment (rholinshead) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:72]: Made a comment above -- we shouldn't pass prompt and we can actually just pass callbacks (and a new `showDeleteOutput` or `canDeleteOutput` prop)

File: python/src/aiconfig/editor/client/src/reducers/aiconfigReducer.ts
Comment (rholinshead) [reply python/src/aiconfig/editor/client/src/reducers/aiconfigReducer.ts:163]: +1, it will handle selecting the correct prompt by id

**Review by rholinshead — CHANGES_REQUESTED** [review]
Overall this looks great. Just requesting changes for some small but important changes in the comments

File: python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx:181]: nit: feel `prompt.outputs?.length > 0` is easier to read and I don't really like JS implicitly "falsey-ness" but don't really care

File: python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:72]: omg this had me confused so hard wondering wtf was going on, but this is the `PromptsContainer`, not `PromptContainer` :S

Comment (rossdanlm) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:72]: I like this change, I feel it makes sense to put with the individual prompt component anyways

File: python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/components/prompt/PromptsContainer.tsx:77]: nice

File: python/src/aiconfig/editor/client/src/shared/types.ts
Comment (rossdanlm) [inline python/src/aiconfig/editor/client/src/shared/types.ts:69]: nit: Would like to use wording "CLEAR_OUTPUT" to match with our UI, but again not a blocker or a big deal

**Review by rossdanlm — APPROVED** [review]
Nice, I think this is great! Will let @rholinshead do a final pass before merging!

File: python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx
Comment (justinkchen) [reply python/src/aiconfig/editor/client/src/components/prompt/PromptContainer.tsx:181]: it actually complains if you use this syntax unelss you do a `(prompt.outputs?.length || 0) > 0` it seems, so thats why i kept it with this way

File: python/src/aiconfig/editor/client/src/shared/types.ts
Comment (justinkchen) [reply python/src/aiconfig/editor/client/src/shared/types.ts:69]: i think maybe as a bigger refactor of the api? i made it clear output at first but then found that all the apis reference delete output so wanted to stay consistent there.

**Review by rholinshead — APPROVED** [review]
Very nice - thanks!

**Comment by rholinshead** [comment]
Ah, looks like the formatting lint is still failing. I'll override to merge to unblock and then I can do a quick PR with the formatting. I'd recommend setting up black formatter extension in vs code (if you use it) so that it formats automatically on save instead of needing the command

**Pull request merged**
